File: NYUDining/DiningLocation.m
//
//  DiningLocation.m
//  NYUDining
//
//  Created by Ross Freeman on 9/8/15.
//  Copyright (c) 2015 Ross Freeman. All rights reserved.
//

#import "DiningLocation.h"

@implementation DiningLocation

-(id)initWithData:(PFObject *)data {
    self = [super init];
    
    if (self) {
        _hours = [[NSMutableArray alloc] initWithCapacity:0];
        _paymentTypes = [[NSMutableArray alloc] initWithCapacity:0];
        _coordinates = [[NSMutableArray alloc] initWithCapacity:0];
        
        
        _hours = [data objectForKey:@"Hours"];
        _name = [data objectForKey:@"Name"];
        _logoURL = [data objectForKey:@"logo_URL"];
        _address = [data objectForKey:@"Address"];
        _phoneNumber = [data objectForKey:@"phone"];
        _coordinates = [[data objectForKey:@"Coordinates"] mutableCopy];
        
        for (NSDictionary *payments in [data objectForKey:@"payments"]) {
            if ([payments objectForKey:@"accepted"]) {
                [_paymentTypes addObject:[payments objectForKey:@"name"]];
            }
        }
        
        
    }
    
    return self;
}


// Determines if location is currently open
-(BOOL)isOpen {
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"EEEE"];
    NSString *dayOfWeek = [dateFormatter stringFromDate:[NSDate date]];
    
    NSDate *now = [NSDate date];
    NSDateFormatter *outputFormatter = [[NSDateFormatter alloc] init];
    [outputFormatter setDateFormat:@"HH"];
    NSString *currentTime = [outputFormatter stringFromDate:now];
    NSInteger currentHour = [currentTime integerValue];
    
    NSInteger timeA = 0;
    NSInteger timeB = 0;

    [outputFormatter setDateFormat:@"a"];
    
    NSString *hoursToday = @"";
    
    if ([dayOfWeek isEqualToString:@"Sunday"]) {
        hoursToday = _hours[0];
    }
    
    else if ([dayOfWeek isEqualToString:@"Monday"]) {
        hoursToday = _hours[1];
    }
    
    else if ([dayOfWeek isEqualToString:@"Tuesday"]) {
        hoursToday = _hours[2];
    }
    
    else if ([dayOfWeek isEqualToString:@"Wednesday"]) {
        hoursToday = _hours[3];
    }
    
    else if ([dayOfWeek isEqualToString:@"Thursday"]) {
        hoursToday = _hours[4];
    }
    
    else if ([dayOfWeek isEqualToString:@"Friday"]) {
        hoursToday = _hours[5];
    }
    
    else if ([dayOfWeek isEqualToString:@"Saturday"]) {
        hoursToday = _hours[6];
    }
    
    if ([hoursToday containsString:@","]) {
        NSArray *timeSlots = [hoursToday componentsSeparatedByString:@","];
        
        NSInteger counter = 0;
        
        for (NSString *timeSlot in timeSlots) {
            
            counter++;
            
            NSArray *times = [timeSlot componentsSeparatedByString:@"-"];
            
            NSMutableString *time0 = [times[0] mutableCopy];
            NSMutableString *time1 = [times[1] mutableCopy];
            
            if ([times[0] containsString:@"am"]) {
                
                if ([times[0] containsString:@"12"]) {
                    timeA = 0;
                }
                
                else {
                    [time0 deleteCharactersInRange:[time0 rangeOfString:@"am"]];
                    timeA = [times[0] integerValue];
                }
            }
            
            else {
                [time0 deleteCharactersInRange:[time0 rangeOfString:@"pm"]];
                
                timeA = [time0 integerValue];
                
                if (timeA < 12) {
                    timeA += 12;
                }
            }
            
            
            
            if ([times[1] containsString:@"am"]) {
                if ([times[1] containsString:@"12"]) {
                    timeB = 0;
                }
                
                else {
                    [time1 deleteCharactersInRange:[time1 rangeOfString:@"am"]];
                    timeB = [times[1] integerValue];
                }
            }
            
            else {
                [time1 deleteCharactersInRange:[time1 rangeOfString:@"pm"]];
                
                timeB = [time1 integerValue];
                
                if (timeB < 12) {
                    timeB += 12;
                }
            }
            
            if (currentHour >= timeA && currentHour < timeB) {
                return YES;
            }
            
            else if (counter == timeSlots.count)
                return NO;
            
        }
        
        
    }
    
    else {
        
        if ([hoursToday isEqualToString:@"Closed"]) {
            return NO;
        }
        else {
            
            NSArray *times = [hoursToday componentsSeparatedByString:@"-"];
            
            NSMutableString *time0 = [times[0] mutableCopy];
            NSMutableString *time1 = [times[1] mutableCopy];
            
            if ([times[0] containsString:@"am"]) {
                
                if ([times[0] containsString:@"12"]) {
                    timeA = 0;
                }
                
                else {
                    [time0 deleteCharactersInRange:[time0 rangeOfString:@"am"]];
                    timeA = [times[0] integerValue];
                }
            }
            
            else {
                [time0 deleteCharactersInRange:[time0 rangeOfString:@"pm"]];
                
                timeA = [time0 integerValue];
                
                if (timeA < 12) {
                    timeA += 12;
                }
            }
            
            
            
            if ([times[1] containsString:@"am"]) {
                if ([times[1] containsString:@"12"]) {
                    timeB = 0;
                }
                
                else {
                    [time1 deleteCharactersInRange:[time1 rangeOfString:@"am"]];
                    timeB = [times[1] integerValue];
                }
            }
            
            else {
                [time1 deleteCharactersInRange:[time1 rangeOfString:@"pm"]];
                
                timeB = [time1 integerValue];
                
                if (timeB < 12) {
                    timeB += 12;
                }
            }
            
            /*
             
             if ([times[0] containsString:@"am"]) {
             
             
             
             [time0 deleteCharactersInRange:[time0 rangeOfString:@"am"]];
             timeA = [times[0] integerValue];
             
             am1 = true;
             }
             
             else if ([times[0] containsString:@"pm"]) {
             
             timeA = [times[0] integerValue];
             
             am1 = false;
             }
             
             if ([times[1] containsString:@"am"]) {
             [time1 deleteCharactersInRange:[time1 rangeOfString:@"am"]];
             timeB = [times[1] integerValue];
             
             am2 = true;
             }
             
             else if ([times[1] containsString:@"pm"]) {
             [time1 deleteCharactersInRange:[time1 rangeOfString:@"pm"]];
             timeB = [times[1] integerValue];
             
             am2 = false;
             }
             
             
             if (am1 && [timeOfDay isEqualToString:@"AM"]) {
             if (currentHour > timeA && currentHour != 12) {
             return YES;
             }
             
             else
             return NO;
             }
             
             else if (am2 && [timeOfDay isEqualToString:@"PM"]) {
             if (currentHour < timeB) {
             return YES;
             }
             
             else
             return NO;
             }
             
             else {
             return NO;
             }
             
             */
            
            if ((currentHour >= timeA && currentHour < timeB) || (currentHour >= timeA && timeB == 0)) {
                return YES;
            }
            
            else
                return NO;
            
        }
    }
    
    return YES;
}

-(BOOL)isOpenTest {
    return NO;
}

-(NSDate *)convertTimeRangeToDate:(NSString *)time {
    
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    NSDate *formattedTime = [formatter dateFromString:time];
    
    NSLog(@"%@", formattedTime);
    
    return formattedTime;
    
    
}


@end

File: NYUDining/DiningLocation.swift
//
//  RFDiningLocation.swift
//  
//
//  Created by Ross Freeman on 6/10/16.
//
//

import Foundation

enum DayOfWeek: String {
    case Sunday = "Sunday"
    case Monday = "Monday"
    case Tuesday = "Tuesday"
    case Wednesday = "Wednesday"
    case Thursday = "Thursday"
    case Friday = "Friday"
    case Saturday = "Saturday"
    case Closed = "Closed"
}

class DiningLocation: NSObject {
    
    let name: String!
    let logoURL: String!
    let menuURL: String?
    let hours: [String]!
    let address: String!
    let coordinates: [Double]!
    let data: [String: AnyObject]
    
    init?(data: [String: AnyObject], params: [String: AnyObject]) {
        self.data = data
        
        guard let cal = params["calendar"] as? String,
            let hours = data[cal] as? [String],
            let name = data["Name"] as? String,
            let logoURL = data["logo_URL"] as? String,
            let address = data["Address"] as? String,
            let coordinates = data["Coordinates"] as? [Double] else {
                return nil
        }
        
        self.hours = hours
        self.name = name
        self.logoURL = logoURL
        self.address = address
        self.menuURL = data["menu_url"] as? String
        self.coordinates = coordinates
        
    }
     
    func isOpen() -> Bool {
        let now = Date()
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "EEEE"
        let dayOfWeek = DayOfWeek(rawValue: dateFormatter.string(from: now))
        
        var hoursToday = ""
        
        if let dayOfWeek = dayOfWeek {
            switch dayOfWeek {
            case .Sunday:
                hoursToday = hours[0]
            case .Monday:
                hoursToday = hours[1]
            case .Tuesday:
                hoursToday = hours[2]
            case .Wednesday:
                hoursToday = hours[3]
            case .Thursday:
                hoursToday = hours[4]
            case .Friday:
                hoursToday = hours[5]
            default:
                hoursToday = hours[6]
            }
        }
        
        if hoursToday.contains(",") {
            let timeComponents = hoursToday.components(separatedBy: ",")
            return isTimeInRange(timeComponents[0]) || isTimeInRange(timeComponents[1])
        }
        
        else {
            return isTimeInRange(hoursToday)
        }
    }
    
    fileprivate func normalizedTime() -> Date {
        let now = Date()
        
        let dateFormatter = DateFormatter()
        dateFormatter.timeStyle = .short
        
        let timeStr = dateFormatter.string(from: now)
        return dateFormatter.date(from: timeStr)!
        
    }
    
    fileprivate func isTimeInRange(_ timeRange: String) -> Bool {
        
        if timeRange == "Closed" {
            return false
        }
        
        let calendar = Calendar.current
        let now = normalizedTime()
        
        let dateFormatter = DateFormatter()
        dateFormatter.timeStyle = .short
        
        let timeComponents = timeRange.components(separatedBy: "-")
        
        let openTime = dateFormatter.date(from: timeComponents[0])
        let closeTime = dateFormatter.date(from: timeComponents[1])
        
        if let openTime = openTime, var closeTime = closeTime {
            
            // Check if close time is 12am
            let diffComparison = (calendar as NSCalendar).compare(openTime, to: closeTime, toUnitGranularity: .hour)
            if diffComparison == .orderedDescending {
                guard let newTime = (calendar as NSCalendar).date(byAdding: .day, value: 1, to: closeTime, options: []) else {
                    return false
                }
                closeTime = newTime
            }
            
            let openComparison = (calendar as NSCalendar).compare(now, to: openTime, toUnitGranularity: .minute)
            let closeComparison = (calendar as NSCalendar).compare(now, to: closeTime, toUnitGranularity: .minute)
            
            return (openComparison == .orderedDescending && closeComparison == .orderedAscending)
        }
        
        return false
    }
    
}
